modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/HeadlessDataSetDisplayTag.java

```diff
@@ -391,15 +394,22 @@ protected void setAttributes(HttpServletRequest httpServletRequest) {
 	}
 
 	private void _setActiveViewSettingsJSON() {
-		PortalPreferences portalPreferences =
-			PortletPreferencesFactoryUtil.getPortalPreferences(request);
+		DatasetCustomViewSettingsFactory datasetCustomViewSettingsFactory =
```

Comment:
Wouldn't this be a breaking change?

There could be existing `portalPreferences` stored, so we might need to either do:
1. An upgrade process
2. An upgrade in place (upgrade a stored setting to a compatible custom view entity) 
3. Just allow this as well as a simplified version of a view.

I think `3. Just allow this` is what makes most sense since many times you might simply make some changes to the dataset and not persist a view, yet you still want that config to be persisted. 

@dsanz, did we take this into account when designing the Custom Views logic?

Reply:
Yes. Mi initial idea was to remove this info as custom views is a new feature, and the amount of things we store in the active view settings is smaller in comparison with the concept of custom view.

Said that, yesterday we talked about a better way to do this, which is basically **using them as per-user defaults**. If there are active view settings, and user did not select any custom view to use, then pre-populate them in the dataset instead of the "default" state hardcoded in the dataset instance. 

To avoid keeping view data outside the view service, when serving the per-user default for the first time, view data should be moved to the custom view service, so that we have a kind of live, auto-upgrade process. 

We could just stop here and we'd get an extra view for that user, which would not differ from any other custom view.

But we might also treat the per-user default view as a bit different as compared to the regular custom view. We could remember any unsaved changes there, so that user would not need to save changes into a named view. This makes things a bit more complex as component has to determine where to save current state (periodically, into the per-user-default view, and when user clicks "save", into the currently active view). 

Doing this would add a nice feature for users, and would do so even if they did not visit the dataset before custom views existed (i.e. there is no need for portal preferences to exist in order to use this feature).


2.

Reply:
I have pushed new commits that implement 2. 

My idea is to always keep a defined migration path possible from what we have now to the new model. After some brainstorming with @dsanz we think we've found a way. He will write it in the specs doc and we can review this PR in light of that migration path.

Reply:
I said I implemented 2, but it is a mix of 2 and three. What I do is always checking if there are associated `portalPreferences` and, in that case, read them and feed them to the new CustomDataViewSettings object. That way, the next time they are saved, they are automatically migrated.

We should, perhaps, delete the `portalPreferences` in that moment too, though another option could be leaving them to be deleted after a reasonable amount of time (for example, on next version, or by a manual process run by the user, ...)